### PR TITLE
Invalidate nodes after graph structure changes

### LINKED ISSUE #1354: Failed actions might be getting stuck in the asset graph
https://gitter.im/dart-lang/angular?at=5ae0b11e2d0e228d7baff65b

If the build fails such that a generated asset node gets marked as a failure, then the build structure changes such that we won't attempt to run that action anymore (but it's still in the graph), we might not be cleaning things up properly.

### PR DESCRIPTION
**Pull request by natebosch**
Fixes #1354

### CODE REVIEW

File: build_runner/lib/src/asset_graph/graph.dart [hunk 1]
```diff
@@ -191,7 +191,9 @@ class AssetGraph {
       var builderOptionsNode = get(node.builderOptionsId);
       builderOptionsNode.outputs.remove(id);
     }
-    _nodesByPackage[id.package].remove(id.path);
+    if (node is! SyntheticSourceAssetNode) {
```
Comment (grouma) [inline build_runner/lib/src/asset_graph/graph.dart]: Add a comment here for why we do this.

File: build_runner/lib/src/asset_graph/graph.dart [hunk 2]
```diff
@@ -309,6 +292,24 @@ class AssetGraph {
         _addOutputsForSources(buildPhases, newIds, rootPackage)
           ..addAll(transitiveRemovedIds);
 
+    // Transitively invalidates all assets.
```
Comment (grouma) [inline build_runner/lib/src/asset_graph/graph.dart]: Optionally add a comment here indicating that this must be done after updates to the graph are completed.

File: build_runner/lib/src/asset_graph/graph.dart [hunk 3]
```diff
@@ -191,7 +191,10 @@ class AssetGraph {
       var builderOptionsNode = get(node.builderOptionsId);
       builderOptionsNode.outputs.remove(id);
     }
-    _nodesByPackage[id.package].remove(id.path);
+    // Synthetic nodes need to be kept to retain dependency tracking.
+    if (node is! SyntheticSourceAssetNode) {
+      _nodesByPackage[id.package].remove(id.path);
```
Comment (jakemac53) [inline build_runner/lib/src/asset_graph/graph.dart]: Note that we are still adding thise node to `removedIds`... is that correct?

Comment (natebosch) [reply build_runner/lib/src/asset_graph/graph.dart]: Good question...

We apparently also explicitly intended for SyntheticSourceAssetNodes to be removed:

https://github.com/dart-lang/build/blob/b7bea8daa7f8fecc866e2b98b20edeb513a6c548/build_runner/lib/src/asset_graph/graph.dart#L449-L451

We count on these removed nodes being tracked so we can remove them from the set of assets consider inputs through a series of phases

https://github.com/dart-lang/build/blob/b7bea8daa7f8fecc866e2b98b20edeb513a6c548/build_runner/lib/src/asset_graph/graph.dart#L419

`allInputs` there starts as the set of sources in the environment, not in the graph, so the synthetic nodes wouldn't be in that set. I _think_ this is safe as is.

**Pull request merged**